File: app/src/main/java/com/example/networkaplication/home/HomeContract.java
package com.example.networkaplication.home;

import android.support.v7.widget.RecyclerView;
import android.text.TextWatcher;
import android.view.View;
import android.widget.TextView;

import com.example.networkaplication.MainActivity;
import com.example.networkaplication.details.DetailsViewFragment;
import com.example.networkaplication.home.adapter.HomeAdapter;
import com.example.networkaplication.home.adapter.ItemData;
import com.example.networkaplication.home.search.story.SearchItem;
import com.example.networkaplication.home.search.story.SearchStoryAdapter;
import com.example.networkaplication.models.search.SearchObject;
import com.example.networkaplication.persistance.model.MovieQuery;

import java.util.ArrayList;
import java.util.List;

public interface HomeContract {

    interface HomeView extends HomeAdapter.OnFilmClickedListener,
            SearchStoryAdapter.OnSearchItemClickedListener, TextView.OnEditorActionListener,
            TextWatcher, View.OnFocusChangeListener {

        MainActivity getHomeViewActivity();

        void setRecycleViewAdapter(HomeAdapter adapter);

        RecyclerView getSearchRecycle();

        void setSearchAdapter(ArrayList<SearchItem> item);

        String getSearchText();

        void setSearchText(String text);

        void setBundleFromSearch(List<SearchObject> objects);

        void setItemDataToAdapter(ArrayList<ItemData> itemData);
    }

    interface  HomePresenter {

        boolean onEditorAction(TextView v, int actionId);

        void setItemToDataAdapter(ArrayList<ItemData> items);

        void refreshSearchAdapter();

        DetailsViewFragment createDetailsView(ItemData itemData);

        void clearSearchAdapter();

        void hideKeyboard(View view);

        void search(String searchRequest);

        void addUniqueMovie(String title);

        void onTextChanged(CharSequence sequence, boolean isSearchPopped);

        void onFocusChanged(View view, boolean hasFocus);

        void onClickedSearchItem(int viewId, SearchItem itemData);
    }

    interface HomeRepository {
        void startSearch(String searchRequest);

        List<MovieQuery> findAllFromQuery(String serach);

        MovieQuery findByTitle(String title);

        void insertIntoQuery(MovieQuery query);
    }
}

File: app/src/main/java/com/example/networkaplication/home/HomeContract.kt
package com.example.networkaplication.home

import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.networkaplication.MainActivity
import com.example.networkaplication.details.DetailsViewFragment
import com.example.networkaplication.home.adapter.HomeAdapter
import com.example.networkaplication.home.adapter.ItemData
import com.example.networkaplication.home.search.story.SearchItem
import com.example.networkaplication.home.search.story.SearchStoryAdapter
import com.example.networkaplication.persistance.model.MovieQuery

interface HomeContract {

    interface HomeView {

        fun getHomeViewActivity(): MainActivity

        fun getSearchRecycle(): RecyclerView

        fun setRecycleViewAdapter(adapter: HomeAdapter)
    }

    interface HomeViewModel : CustomOnEditorActionListener, CustomOnFocusChangeListener, SearchStoryAdapter.OnSearchItemClickedListener, HomeAdapter.OnFilmClickedListener {

        override fun onEditorAction(v: TextView?, actionId: Int)

        fun setItemToDataAdapter()

        fun refreshSearchAdapter()

        fun createDetailsView(itemData: ItemData): DetailsViewFragment

        fun clearSearchAdapter()

        fun hideKeyboard(view: View)

        fun search(searchRequest: String)

        fun addUniqueMovie(title: String)

        fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int)

        override fun onFocusChanged(view: View, hasFocus: Boolean)

        override fun onClickedSearchItem(viewId: Int, itemData: SearchItem)

        fun isSearchItemEmpty(): Boolean
    }

    interface HomeRepository {
        fun startSearch(searchRequest: String)

        fun findAllFromQuery(search: String): List<MovieQuery>

        fun findByTitle(title: String): MovieQuery?

        fun insertIntoQuery(query: MovieQuery)
    }
}
